---
title:  "CHIKV 28dpi scRNA-seq"
author: "Ryan Sheridan"
date:   "`r Sys.Date()`"
output: 
  html_document:
    toc:       true
    toc_float: true
    toc_depth: 2
    theme:     cosmo
    highlight: kate
params:
  
  unenr_dir:    "results/2021-11-05"      # Directory for unenriched data
  enr_dir:      "results/2022-02-03"      # Directory for CHIKV-enriched data
  template_dir: "src"                     # Directory for Rmd templates
  ref_dir:      "ref"
  table_dir:    "results/tables/2024-03-29"          # Directory to write tables
  geo_dir:      "results/geo"             # Directory for GEO submission
  overwrite:    false                     # Should tables be overwritten if they already exist?
  so_dir:       "~/Dropbox/Ryan/Projects/morrison-chronic-chikv/results/sobjs/2024-01-22_4" # Directory for Seurat objects
  metrics:     "count_metrics.csv"       # Cell Ranger metrics
  gene_min:    250                       # Min number of detected genes per cell
  gene_max:    5500                      # Max number of detected genes per cell
  mito_max:    30                        # Max percentage mito reads per cells
  rm_doublets: true                      # Remove doublets
  type_res:    5                         # Resolution for annotating cell types
  mac_res:     2                         # Resolution for annotating macrophages
  t_res:       1                         # Resolution for annotating T cells
  fib_res:     5                         # Resolution for annotating fibroblasts cells
  tm:          "28dpi"                   # Label for time point
  n_threads:   2
  rslns:
    value:
      type: 5
      mac:  2
      t:    1
      fib:  5
  samples:
    - "U1"
    - "U2"
    - "U3"
    - "C1"
    - "C2"
    - "C3"
  treatment_levels: ["mock", "CHIKV"]  # order as control, treatment, order affects how marker genes are identified
editor_options:
  chunk_output_type: console
---

<br>

## Summary

* Genes from the combined Xenium panel (mouse atlas + custom genes) were
  assessed for expression in the CHIKV 28 dpi datasets.
* Markers are present in the Xenium panel that can be used to identify
  macrophages.
* Some genes upregulated in macrophages from CHIKV-infected mice are also
  present in the panel.

---

<br>

<br>

```{r "setup", include = FALSE}
# Default chunk options
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo    = FALSE,
  dpi     = 300
)

options(future.globals.maxSize = 5.5 * 1024^3)

knitr::knit(here::here("src/setup.Rmd"), output = "")

# Columns to exclude when saving DE tables
exclude_clmns <- c(
  "max_pval", "minimump_p_val", "avg_fc",
  "all_same_direction", "n_same_direction",
  "significant"
)

# Function to create markers plots
.create_marker_fig <- function(dat, gns, clrs, lvls = NULL, n_gns = 20) {
  
  gns <- head(gns, n_gns)
  
  dat <- dat %>%
    filter(name %in% gns) %>%
    mutate(
      name      = fct_relevel(name, gns),
      cell_type = fct_reorder(cell_type, value, .fun = mean, .desc = TRUE),
      cell_type = fct_relevel(cell_type, lvls)
    )
  
  n_labs <- dat %>%
    group_by(cell_type) %>%
    summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
    mutate(
      n = if_else(n > 1000, str_c(round(n / 1000, 1), "k"), as.character(n)),
      n_lab = str_c(cell_type, " (", n, ")")
    )
  
  n_labs <- set_names(n_labs$n_lab, n_labs$cell_type)
  
  res <- gns %>%
    imap(~ {
      p <- dat %>%
        filter(name == .x) %>%
        ggplot(aes(cell_type, value, fill = cell_type)) +
        geom_boxplot(outlier.size = 0.01, outlier.alpha = 0.1) +
        facet_wrap(~ name) +
        scale_fill_manual(values = clrs) +
        scale_x_discrete(labels = n_labs) +
        labs(y = "expression") +
        umap_theme_2 +
        theme(
          legend.position = "none",
          strip.clip      = "off",
          axis.title.y    = element_text(),
          axis.ticks.x    = element_line(color = ln_col),
          axis.text.x     = element_text(size = 8, angle = 90, hjust = 1, vjust = 0.5),
          axis.text.y     = element_text(size = 8)
        )
      
      if (.y %% 5 != 1) {
        p <- p + 
          theme(axis.title.y = element_blank())
      }
      
      if (length(gns) > 5 && .y <= (floor(max(length(gns) - 1, 1) / 5) * 5)) {
        p <- p +
          theme(
            axis.text.x  = element_blank(),
            axis.ticks.x = element_blank()
          )
      }
      
      p
    }) %>%
    wrap_plots(
      nrow = 4,
      ncol = 5,
      guides = "collect"
    )
  
  res
}
```

```{r "xenium panels"}
xen_sheet <- "/home/rmsheridan/Projects/xenium-probe-design/xenium-panel.xlsx"

xen_probes <- xen_sheet %>%
  read.xlsx(sheetName = "final-panel-1") %>%
  as_tibble()

mouse_atlas <- xen_sheet %>%
  read.xlsx(sheetName = "Xenium mouse atlas") %>%
  as_tibble() %>%
  select(-1)

xen_genes <- unique(c(mouse_atlas$gene, xen_probes$gene))

xen_genes <- xen_genes[xen_genes %in% rownames(so)]
```

## Mock {.tabset .tabset-pills}

Mean expression for genes included in the Xenium panel is shown below
for each cell type.

```{r "xenium heatmaps", fig.height = 5}
xen_dat <- so %>%
  FetchData(c("treatment", "cell_type", xen_genes), slot = "data") %>%
  as_tibble(rownames = ".cell_id") %>%
  pivot_longer(any_of(xen_genes)) %>%
  group_by(treatment, cell_type, name) %>%
  summarize(value = mean(value), .groups = "drop") %>%
  group_by(treatment, name) %>%
  filter(sum(value) > 0) %>%
  mutate(value = as.numeric(scale(value))) %>%
  ungroup()

xen_mat <- xen_dat %>%
  split(.$treatment) %>%
  map(~ {
    .x %>%
      pivot_wider(names_from = name, values_from = value) %>%
      select(-treatment) %>%
      column_to_rownames("cell_type") %>%
      t()
  })

heats <- xen_mat %>%
  imap(~ {
    .x %>%
      ComplexHeatmap::Heatmap(
        show_row_names = FALSE,
        col = c("lightblue", "white", "red"),
        name = str_c(.y, "\nmean\nexpression")
      )
  })

heats$mock
```

<br>

Expression of the top 20 marker genes present in the Xenium panel is shown below
for each cell type.

```{r "xenium cell type markers"}
Idents(so) <- so$cell_type

mkrs <- treats %>%
  set_names() %>%
  map(~ {
    mks <- so %>%
      subset(treatment == .x) %>%
      FindAllMarkers(
        features = xen_genes,
        only.pos = TRUE
      )
    
    mks %>%
      filter(pct.1 > 0.25) %>%
      group_by(gene) %>%
      mutate(n_types = n_distinct(cluster)) %>%
      ungroup() %>%
      arrange(cluster, n_types, desc(p_val_adj), desc(avg_log2FC))
  })

gn_dat <- so %>%
  FetchData(c("treatment", "rep", "cell_type", xen_genes)) %>%
  as_tibble(rownames = ".cell_id") %>%
  pivot_longer(all_of(xen_genes)) %>%
  split(.$treatment)
```

```{r, fig.width = 9, fig.height = 9, results = "asis"}
trt <- "mock"

typs <- mkrs[[trt]]$cluster %>%
  table() %>%
  sort(decreasing = TRUE) %>%
  names()

typs %>%
  walk(~ {
    cat("\n\n###", .x, "\n\n")
    
    gns <- mkrs[[trt]] %>%
      filter(cluster == .x) %>%
      pull(gene)
    
    plt <- .create_marker_fig(
      dat  = gn_dat[[trt]],
      gns  = gns,
      clrs = type_cols,
      lvls = .x
    )
    
    print(plt)
    
    cat("\n\n---\n\n<br>\n\n<br>\n\n")
  })
```

## CHIKV-infected {.tabset .tabset-pills}

Mean expression for genes included in the Xenium panel is shown below
for each cell type.

```{r "xenium heatmaps-2", fig.height = 5}
heats$CHIKV
```

<br>

Expression of the top 20 marker genes present in the Xenium panel is shown below
for each cell type.

```{r, fig.width = 9, fig.height = 9, results = "asis"}
trt <- "CHIKV"

typs <- mkrs[[trt]]$cluster %>%
  table() %>%
  sort(decreasing = TRUE) %>%
  names()

typs %>%
  walk(~ {
    cat("\n\n###", .x, "\n\n")
    
    gns <- mkrs[[trt]] %>%
      filter(cluster == .x) %>%
      pull(gene)
    
    plt <- .create_marker_fig(
      dat  = gn_dat[[trt]],
      gns  = gns,
      clrs = type_cols,
      lvls = .x
    )
    
    print(plt)
    
    cat("\n\n---\n\n<br>\n\n<br>\n\n")
  })
```

## Mock vs CHIKV macrophages

Expression of genes upregulated in macrophages from CHIKV-infected mice that are
also present in the Xenium panel is shown below.

```{r, "mock vs CHIKV markers", fig.height = 7, fig.width = 8}
# Read CHIKV macrophage markers
chikv_mkrs <- "/home/rmsheridan/Projects/morrison-chronic-chikv/results/tables/2024-03-29/celltype_chikv_markers.tsv.gz"

chikv_mkrs <- chikv_mkrs %>%
  read_tsv()

chikv_mkrs <- chikv_mkrs %>%
  filter(
    gene %in% xen_genes,
    direction == "up",
    significant
  ) %>%
  arrange(max_p_adj)

# Format data for plotting
mac_gns <- chikv_mkrs %>%
  filter(cell_type == "Macrophages") %>%
  pull(gene)

mac_dat <- bind_rows(gn_dat) %>%
  mutate(
    treatment = fct_relevel(treatment, treats),
    name      = fct_relevel(name, mac_gns)
  ) %>%
  filter(
    name %in% mac_gns,
    cell_type == "Macrophages"
  )

# Create boxplots
n_labs <- mac_dat %>%
  group_by(treatment) %>%
  summarize(n = n_distinct(.cell_id), .groups = "drop") %>%
  mutate(
    n = if_else(n > 1000, str_c(round(n / 1000, 1), "k"), as.character(n)),
    n_lab = str_c(treatment, " (", n, ")")
  )
  
n_labs <- set_names(n_labs$n_lab, n_labs$treatment)

mac_dat %>%
  ggplot(aes(treatment, value, fill = treatment, alpha = rep)) +
  geom_boxplot(outlier.size = 0.01, outlier.alpha = 0.1) +
  facet_wrap(~ name, ncol = 7, scales = "free_y") +
  scale_fill_manual(values = c(mock = "#56B4E9", CHIKV = "#D7301F")) +
  scale_x_discrete(labels = n_labs) +
  labs(y = "expression") +
  base_theme +
  theme(
    aspect.ratio = 1,
    legend.position = "none",
    legend.title = element_blank(),
    panel.border = element_rect(color = ln_col, fill = NA),
    strip.clip   = "off",
    axis.text.x  = element_text(angle = 90, hjust = 1, vjust = 0.5),
    axis.text.y  = element_text(size = 6),
    axis.title.x = element_blank()
  )
```

<br>

<br>

## Session info

```{r}
sessionInfo()
```
